### ISSUE TITLE
The call is ambiguous when using collections literals with interpolated strings

### ISSUE THREAD

**Issue by xamir82** [dotnet/roslyn#70238]
The following code doesn't compile:
```csharp
MyClass.MyMethod([$"one", $"two"]);

public static class MyClass
{
    public static void MyMethod(IEnumerable<string> _)
    {

    }
    public static void MyMethod(IEnumerable<IFormattable> _)
    {

    }
}
```

If we had used `new[] { $"one", $"two" }` it would work. I would expect collection literals to work the same way as `new[] { }` here. Let me emphasize that this only errs when you're using interpolated strings, it's kind of weird.

Sharplab: https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQAYACdbAFgG4CCBZATwGEAbCAZ3YDpG6BTAC4ALAPYIAFAG0AJACIRMPrLgk5AgO4jZAXQCUNfAXQBmMtgBsZTCUasO7AgG8CJV2VPlL6CjYb9hYuLkxgA8AJIAYiIATpACAhDALHwAfCQA+roubs6G+G4kAL60+W4mZl4+vIKiEsEh5ERpmdmuua1FBIVAA

**Comment by xamir82** [comment]
> I'm personally ok with this not working

But this pair of overloads is not uncommon, and therefore not being able to use collection expressions in this context is kind of a bummer. One example is the popular [CliWrap](https://github.com/Tyrrrz/CliWrap) library, where the `args.Add` method has these exact overloads; which means this errs:

```csharp
Cli.Wrap("command").WithArguments(args =>
{
    args.Add(["--some-flag", $"some-value-{whatever}");
});
```

**Comment by xamir82** [comment]
Why does this work by the way?

```csharp

MyClass.Foo($"");

public static class MyClass
{
    public static void Foo(IFormattable _)
    {

    }

    public static void Foo(string _)
    {

    }
}
```
It picks the string overload, as expected. Why can't collection expressions behave the same way?!

**Comment by CyrusNajmabadi** [comment]
> It picks the string overload, as expected. So, there's currently an asymmetry; why couldn't collection expressions do the same?!

It could potentially could.  If there's a betterness rule between interpolated strings and strings vs iformattables, then that possibly should apply in terms of determining betterness here.  I'm curious what @333fred thinks.

**Comment by RikkiGibson** [comment]
```cs
using System;

class Program
{
    void M0(IFormattable format) { }
    void M0(string format) { }

    void M1() { M0($"a"); }
    
    
    void M2(IFormattable[] format) { }
    void M2(string[] format) { }

    void M3() { M2([$"a"]); }
}
```

In the above sample I basically expect the call to M2 to work because the call to M0 works. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoTyAECUDsmA3pjuXgCw4CyADABQCSAYgPYBOkALtxMABsApjgBmnHgEocxHAF8yFBNXoN8dMRIjdpshVgwUqtFA121GAEgBEEa5IDc8xeRc43y2kmbsu2voJCANoAupp+OjLOhkoq3uqh4VJR+h4qAMxmUTTeQTZ2IY7RckA=)

**Comment by xamir82** [comment]
@RikkiGibson Yeah exactly, that's what I was saying

**Comment by 333fred** [comment]
This is something that I somewhat expect we will need natural typing to solve. `new[] { $"" }` works here for a simple reason: it isn't target-typed. If you look at the [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQAYACdbAFgG4CCBZATwGEAbCAZ3YDpG6BTAC4ALAPYIAFDD4B3ANoBdEgG8AJACIRUtXBLqB0kWoC+AShr4C6AMxlsANjKYSjVh3YElBEt7I3yD9ApnBn5hMXFyKwAeAEkAMREAJ0gBAQhgFj4APhIAfRMvH08LfB8SIwIjIA===) when removing the `IEnumerable<string>` overload for that case, you'll find that `new[] { $"" }` cannot be converted to `IEnumerable<IFormattable>` or `IEnumerable<FormattableString>`. The natural type of the array is `string[]`, and that is simply not convertible to `IEnumerable<IFormattable>`. Overload resolution doesn't even come into play: there's only one applicable `M` in this example for `new[] { $"" }`.

`[$""]`, on the other hand, _is_ target-typed. You can pass that to both `M(IEnumerable<string>)`, and `M(IEnumerable<IFormattable>)`. Overload resolution therefore needs to come into play, and you cannot convert `IEnumerable<IFormattable>` to `IEnumerable<string>` or vice versa. Therefore, it's ambiguous. When collection expressions have a natural type, we may be able to then say, in the face of an ambiguous overload, if one is a natural type match, that should win out.

---

Looking at the `M0` example Rikki posted, there's a betterness rule that applies here: which conversion is better. `$"a"` (which I will refer to as expression E) _is_ a `string`. It has an identity conversion. E to `IFormattable` is an implicit interpolated string conversion. Because the type of E is `string`, that is the better conversion, so it wins in betterness. There is no nested conversion comparisons in betterness there; the algorithm considers top-level betterness only.

---

To sum this up, this is (imo) expected behavior, and will need to wait until C# 13 to be addressed.

**Comment by CyrusNajmabadi** [comment]
@333fred  that makes perfect sense to me.  We'll put this on the list of things to try to ensure works in C# 13.

**Comment by RikkiGibson** [comment]
Triaging to 17.9 as we are no longer using Compiler.Next etc milestones.

**Comment by 333fred** [comment]
@RikkiGibson I don't think 17.9 is right. I don't expect that we can address this in C# 12.

**Comment by xamir82** [comment]
> When collection expressions have a natural type, we may be able to then say, in the face of an ambiguous overload, if one is a natural type match, that should win out.

Isn't the natural type related to the collection type? e.g. `List<string>` vs `string[]` etc.? How would that solve this problem?

**Comment by 333fred** [comment]
> Isn't the natural type related to the collection type? e.g. List<string> vs string[] etc.? How would that solve this problem?

Yes, it is, and that's exactly why it would solve this case. `List<string>` is convertible to `IEnumerable<string>`, but not to `IEnumerable<IFormattable>`, so the former would win. Now, if the answer you were hoping for was that it would pick the `IEnumerable<IFormattable>` overload... well, that may need more specific spec changes to enable.

**Comment by xamir82** [comment]
> Yes, it is, and that's exactly why it would solve this case. List<string> is convertible to IEnumerable<string>, but not to IEnumerable<IFormattable>, so the former would win. Now, if the answer you were hoping for was that it would pick the IEnumerable<IFormattable> overload... well, that may need more specific spec changes to enable.

No I was hoping for `IEnumerable<string>` so that's fine, I just want it to behave the same way as the singular `string` vs `IFormattable` overloads.

But what happens if you decide that the natural type is `List<T>` and the overloads receive `string[]` and `IFormattable[]`? Would the taget-typed algorithm kick in again in that case, leading to the same issue?

**Comment by 333fred** [comment]
> But what happens if you decide that the natural type is List<T> and the overloads receive string[] and IFormattable[]? Would the taget-typed algorithm kick in again in that case, leading to the same issue?

Potentially, yes. I do think the working group needs to consider this case @CyrusNajmabadi @jnm2 @captainsafia @cston. I don't have any immediate ideas for how best to solve it, but I do think that we need to. The closest analogy I can find here is the tuple case, which does work:

```cs
using System;

MyClass.MyMethod(($"one", $"two")); // Calls the string tuple overload

public static class MyClass
{
    public static void MyMethod((string, string) _)
    {

    }
    public static void MyMethod((IFormattable, IFormattable) _)
    {

    }
}
```

Given that we've previously said we want to consider the tuple behavior the model for how we want collection expressions to work, it does feel like we're missing a rule somewhere here.

**Comment by jnm2** [comment]
The tuple case was exactly what I was going to suggest we follow. Can we glean rules straight across perhaps?

**Comment by xamir82** [comment]
> Potentially, yes.

Hmm okay that doesn't seem like the optimal solution then.

> Given that we've previously said we want to consider the tuple behavior the model for how we want collection expressions to work

That's the right path IMO, and again, I would argue this should _really_ be done in v12. You'd very much intuitively expect it to work.

**Comment by RikkiGibson** [comment]
> @RikkiGibson I don't think 17.9 is right. I don't expect that we can address this in C# 12.

No worries, feel free to re-triage as appropriate.

I think the issue with doing it for 12 is time. There's so much we're trying to get in for this feature in 12 right now with a rapidly shrinking window of opportunity to do so.

**Comment by aradalvand** [comment]
I obviously have no clue about the inner workings of the compiler but given that the logic for this has already been implemented for tuples as mentioned earlier, it should not be too time-consuming to bring it over, but correct me if I'm wrong.

I agree that the currently behavior is definitely surprising.

**Comment by 333fred** [comment]
> it should not be too time-consuming to just adapt it to collection expressions before the C# 12 release

This is never a safe assumption to make :laughing:

**Comment by aradalvand** [comment]
> This is never a safe assumption to make 😆

Fair enough 😬😅 But hopefully you give it a try.

**Comment by 333fred** [comment]
> But hopefully you give it a try.

It isn't going to happen. The cutoff date for such changes is basically today. It will be a post-release thing.

**Comment by xamir82** [comment]
> It will be a post-release thing.

Does that mean we need to wait until C# 13 or will this be done before that?

**Comment by CyrusNajmabadi** [comment]
> Given that we've previously said we want to consider the tuple behavior the model for how we want collection expressions to work, it does feel like we're missing a rule somewhere here.

@333fred @cston could you look into why this works for tuples but not collections? That does feel strange.  Could it just be an implementation bug where were doing something for tuples, and forgetting to do the same here?

**Comment by CyrusNajmabadi** [comment]
@333fred @cston .  Something else surprising:

```c#
MyClass.MyMethod($"one", $"two");

public static class MyClass
{
    public static void MyMethod(params string[] _)
    {

    }
    public static void MyMethod(params IFormattable[] _)
    {

    }
}
```

Note the use of 'params' and no collection-expression.  Here we *do* expand things out, and find that the string overload is the preferred one.  This also breaks the intuition that a params-arrays and a collection-expr are nearly the same (except for brackets).

I'm guessing/hoping, that there's just somethgin small being missed here.  Maybe not calling through a common 'betterness' helper versus just writing the rules out explicitly?

**Comment by xamir82** [comment]
@333fred @cston Could you please respond to @CyrusNajmabadi's comment?

**Comment by CyrusNajmabadi** [comment]
@xamir82 Please give people time to respond.  Everyone has enormously full plates right now and are working on many things at the same time.  We care about this issue and will respond when there is time.  Thanks! :)

**Comment by cston** [comment]
For overload resolution, *better conversion from expression* prefers the type where the expression matches exactly - see [*exactly-matching-expression*](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/expressions.md#12646-exactly-matching-expression).

For the tuple example, the expression `($"one", $"two")` has a *natural type* of `(string, string)` which exactly matches one of the two overloads, so that overload is preferred.

Note however, the tuple `($"one", null)` does *not* have a natural type, so that case is *ambiguous* - see [sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBQLICeAwgDYQDOzAdLVQKYIAWA9hgAUQgCQAifjC7i4AAgkIA7v3EBKNQG45Aeh1y6EBg2Zy+XOfmIAGM6gAODC/wBuXZA34QM1ek1YcNNx8giISUjLyMKjGGtpyuvoQYMBQAOao/KjMFBT4AMyWxABsloRytIwsOeQA3hQJCQVFpfgALBVBPALCQlbW8v1qcgD6ag2N9ZTkjXIAvhNNhVatHZzdoUIAkgBi/MiQCAgQwE7yu/uHx6dcw2OLclMPC+RzQA==).

For collection expressions, perhaps [*exactly-matching-expression*](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/expressions.md#12646-exactly-matching-expression) should recurse into the collection expression and check if the elements exactly match the element type of the target type. Similar to the tuple examples, we'd need to decide whether elements without types, such as `null` in `[$"one", null]` are considered.

**Comment by CyrusNajmabadi** [comment]
@cston  We probably need to do this.  With tuples, it's less of an issue as there's only 'one possible tuple' you can ever get.  So the dependence on natural type is ok.  But with collection expressions our natural type will be something like `List<T>`.  But that won't work properly for all the different collection types a user might have.  for example, we'd want consistent behavior if you had:

```c#
public static void MyMethod(IEnumerable<string> _)
public static void MyMethod(IEnumerable<IFormattable> _)

// or

public static void MyMethod(string[] _)
public static void MyMethod(IFormattable[] _)

// or

public static void MyMethod(Span<string> _)
public static void MyMethod(Span<IFormattable> _)
```

etc. :)

**Comment by jaredpar** [comment]
@cston 

> For the tuple example, the expression ($"one", $"two") has a natural type of (string, string) so that expression exactly matches one of the two overloads, which is preferred.

My take away from this is that fixing this requires us to have natural type support in collection expressions. Is that correct?

**Comment by aradalvand** [comment]
@jaredpar I think this was already discussed, and the conclusion was that no, even with a natural type, this [would still be problematic](https://github.com/dotnet/roslyn/issues/70238#issuecomment-1747714032).

**Comment by cston** [comment]
> My take away from this is that fixing this requires us to have natural type support in collection expressions. Is that correct?

If we use natural type, we'd probably want a natural *element type* only, to avoid ambiguities when the overloads use other *collection types*.

**Comment by jaredpar** [comment]
Okay given this is _natural type_ related that means it's not in scope for C# 12. This would be part of the C# 13 design of collection expressions improvements. It's effectively "by design" for the current version. As such going to close this issue out. 

Will address it when we approach the C# 13 design.

**Comment by xamir82** [comment]
@jaredpar But why would you close the issue?!

**Comment by jnm2** [comment]
I like the idea of transfer.

**Comment by xamir82** [comment]
> This instead effectively a request to change the design of the feature. It's a better discussion for csharplang.

I don't think so, I don't really understand how drew that conclusion, this _is_ an implementation bug.

**Comment by jnm2** [comment]
It's not an implementation bug when it's the spec that would need design work. Spec bugs are discussed at csharplang, compiler-not-matching-spec bugs are discussed here.
